### PR TITLE
Implement `FixedSizeArray` for all fixed size arrays

### PR DESCRIPTION
**Pull request by tbu-**

### CODE REVIEW

**Comment by tbu-** [comment]
Code from [this reddit post](https://www.reddit.com/r/rust/comments/3irc91//cuj43qq) by @eddyb.

**Comment by bluss** [comment]
This is quite amazing. Ideally the element type should be an associated type, but I know the impl isn't allowed that way.

File: src/libcore/array.rs [hunk 1]
```diff
@@ -41,21 +41,21 @@ pub trait FixedSizeArray<T> {
     fn as_mut_slice(&mut self) -> &mut [T];
 }
 
+impl<T, A: Unsize<[T]>> FixedSizeArray<T> for A {
+    #[inline]
+    fn as_slice(&self) -> &[T] {
+        unsafe { &*(self as *const A as *const [T]) }
```
Comment (bluss) [inline src/libcore/array.rs]: This can be just `self as &[T]` or in fact just `self` I think, and let coercion do the rest?

Comment (heinrich5991) [inline src/libcore/array.rs]: Indeed. How nice. :)

**Comment by huonw** [comment]
(It'd be great to have a more descriptive commit message or maybe some comments: I can definitely see people looking at this and wondering what's going on in future.)

**Comment by eddyb** [comment]
How well does it work with coherence?
Can you, e.g. implement `Clone` for `[T; N]` via `A: Unsize<[T]>` without breaking anything?

**Comment by tbu-** [comment]
@eddyb No, it unfortunately doesn't work with coherence at all.

**Comment by eddyb** [comment]
@tbu- I tried making it smarter at some point: it should be possible to teach coherence that `impl<T> Clone Foo<T>` cannot conflict with `impl<T, A: Unsize<[T]>> Clone for A` due to `Foo<_>: Unsize<[_]>` never holding, but it's non-trivial and quite hacky.
cc @nikomatsakis @arielb1 Maybe there's a better way to handle this.

**Comment by alexcrichton** [comment]
Using this for other blanket impls would make me a little uneasy, but this is an unstable trait anyway so I think this is fine. Could you squash the commits together and expand the first commit message like @huonw mentioned? Otherwise looks good to me!

**Comment by tbu-** [comment]
@alexcrichton Squashed and expanded the first commit message.

**Comment by bluss** [comment]
Yeah I don't believe this is what we'd want in the future (it's not how trait support for fixed size arrays would look if we had designed for it).

**Comment by petrochenkov** [comment]
Does anybody actually use `FixedSizeArray`?
It couldn't be used for impls like
```
impl<U: Clone> Clone for T where T: FixedSizeArray<U> {...}
```
like it was [initially planned](https://github.com/rust-lang/rust/pull/22838#issuecomment-82489475) due to coherence restrictions, and without it it's almost useless, probably I should have been kill it in the same PR that introduced it.

**Comment by tbu-** [comment]
@petrochenkov The trait is useful for abstractions over arrays, such as stack-allocated vectors, hybrid vectors, etc.

**Comment by tbu-** [comment]
@alexcrichton I believe there's a space missing after the colon in `@bors: retry`, this pull request isn't in approved state while #28115 is after `@bors: retry`.

**Pull request merged**